In [41]:
# 1 Perform a set of exercises. All tasks should be performed in two ways:
#(A)using universal functions of the NumPy library
import numpy as np

angles = np.array([1, 3, 4, 6,10, 12, 15,17, 18])


radians = np.deg2rad(angles)

print('Sine of angles in the array:')
sine_value = np.sin(radians)
print(np.sin(radians))

print('Inverse Sine of sine values:')
print(np.rad2deg(np.arcsin(sine_value)))

print('Sine hyperbolic of angles in the array:')
sineh_value = np.sinh(radians)
print(np.sinh(radians))

print('Inverse Sine hyperbolic:')
print(np.sin(sineh_value))

base = 4
height = 3
print('hypotenuse of right triangle is:')
print(np.hypot(base, height))


Sine of angles in the array:
[0.01745241 0.05233596 0.06975647 0.10452846 0.17364818 0.20791169
 0.25881905 0.2923717  0.30901699]
Inverse Sine of sine values:
[ 1.  3.  4.  6. 10. 12. 15. 17. 18.]
Sine hyperbolic of angles in the array:
[0.01745418 0.05238381 0.06986989 0.10491126 0.17542037 0.21097405
 0.26480023 0.30107857 0.31935254]
Inverse Sine hyperbolic:
[0.01745329 0.05235985 0.06981306 0.10471891 0.17452207 0.20941245
 0.26171646 0.29655043 0.3139519 ]
hypotenuse of right triangle is:
5.0


In [44]:
#(A)calculate the execution time of the script.
%timeit (1.0 / big_array)

4 ms ± 117 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [42]:
# 1 Perform a set of exercises. All tasks should be performed in two ways:
#(B)using iterative constructions (using loops, list comprehensions, etc.
a = np.arange(18)

for x in np.nditer(a, flags=['external_loop']):
    print(x, end=' ')
    

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17] 

In [45]:
#(B)calculate the execution time of the script
%timeit (1.0 / big_array)

3.88 ms ± 164 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# 1. Perform a set of exercises. All tasks should be performed in two ways:
Time in Task 1(A) 4 ms ± 117 µs per loop (mean ± std. dev. of 7 runs, 100 loops each), 
Time in Task 1(B) 3.88 ms ± 164 µs per loop (mean ± std. dev. of 7 runs, 100 loops each), 
as seen loop (B) is slightly fast than loop (A) 

In [53]:
#2.Solve the system of linear equations using Cramer's formulas and check the result using:
#(A) matrix multiplication;
import numpy as np

a = np.array([[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]])
b= np.array([20, 9, -9, 1])

print("A =", a)

print("b =", b)

print("Ab =",np.matmul(a,b))


A = [[ 1  5  3 -4]
 [ 3  1 -2  0]
 [ 5 -7  0 10]
 [ 0  3 -5  0]]
b = [20  9 -9  1]
Ab = [34 87 47 72]


In [51]:
#2.Solve the system of linear equations using Cramer's formulas and check the result using:
#(B) inverse matrix;
import numpy as np

m_list = [[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]]
A = np.array(m_list)

inv_A = np.linalg.inv(A)

print(inv_A)

B = np.array([20, 9, -9, 1])
X = np.linalg.inv(A).dot(B)

print(X)

[[ 0.01587302  0.31746032  0.00634921 -0.11746032]
 [ 0.23809524 -0.23809524  0.0952381   0.23809524]
 [ 0.14285714 -0.14285714  0.05714286 -0.05714286]
 [ 0.15873016 -0.32539683  0.16349206  0.22539683]]
[ 3.  2.  1. -1.]


In [57]:
#2.Solve the system of linear equations using Cramer's formulas and check the result using:
#(C) numpy.linalg.solve().
import numpy as np

a = np.array([[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]])
b = np.array([20, 9, -9, 1])
x = np.linalg.solve(a, b)
x

array([ 3.,  2.,  1., -1.])

In [63]:
# 2 Compare the results using numpy.allclose().
np.allclose([34, 87, 47, 72], [3,  2,  1, -1], [3,  2,  1, -1])

False

In [109]:
# 3 Calculate the value of the matrix expression
# (A) using the universal functions of the NumPy library.

import numpy as np

a = np.array([5, 2, 0, 10, 4, 1, 7, 3, 2])
b = np.array([3, 6, -1, -1, -2, 0, 2, 1, 3])

Bb= np.array(b ** 2)
aa= np.array(a * 2)

print("Value =", (a - Bb)*(aa + b))

Value = [ -52 -340    1  171    0    2   48   14  -49]


In [100]:
#(A)calculate the execution time of the script.
%timeit (1.0 / big_array)

4.1 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [118]:
# 3 Calculate the value of the matrix expression
# (B) using iterative constructions (using loops, list comprehensions, etc.).

A = [[5, 2, 0], [10, 4, 1], [7, 3, 2]]
B = [[3, 6, -1], [-1, -2, 0], [2, 1, 3]]

result = [[sum(a*b for a,b in zip(A_row,B_col)) for B_col in zip(*B)] for A_row in A]

for r in result:
    print(r)

[13, 26, -5]
[28, 53, -7]
[22, 38, -1]


In [119]:
#(B)calculate the execution time of the script.
%timeit (1.0 / big_array)

3.96 ms ± 143 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# 3. Calculate the value of the matrix expression:
Time in Task 3(A) 4.1 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each), 
Time in Task 3(B) 3.96 ms ± 143 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
as shown once again iterative constructions are slightly faster can ufunc in processing speed

:D